File: leetcode/287-Find-the-Duplicate-Number.swift
class Solution {
    func findDuplicate(_ nums: [Int]) -> Int {
        var fast = nums[0]
        var slow = nums[0]

        repeat {
            fast = nums[nums[fast]]
            slow = nums[slow]
        } while fast != slow

        fast = nums[0]

        while fast != slow {
            fast = nums[fast]
            slow = nums[slow]
        }

        return fast
    }
}

File: Algorithms/287-Find-the-Duplicate-Number.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        q, s = nums[0], nums[0]
        while True:
            q = nums[nums[q]]
            s = nums[s]
            if q == s:
                break
        q = nums[0]
        while q != s:
            q = nums[q]
            s = nums[s]
        return q